### PR TITLE
Watcher/general metrics

### PR DESCRIPTION
**Pull request by kjdelisle**
Cherry-picked from:
- https://github.com/strongloop/strong-supervisor/commit/98b8d2051f1b547e0938195aea4766d9fd7df47e
- https://github.com/strongloop/strong-supervisor/commit/ee3e51edabc01e2b24413550d5bb7a39ebc67620

Additional linting fixes included.

### CODE REVIEW

File: lib/watcher/appmetrics.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  'redis',
+  'memcached',
+  'oracledb',
+  'strong-oracle',
+  'requests',
+  'trace',
+];
+
+exports.worker = function(handle) {
+  var appmetrics = require('appmetrics');
```
Comment (sam-github) [inline lib/watcher/appmetrics.js]: you have to find a way to use `../adapter`, so that `sl-run --no-profile` causes no appmetrics to be required

File: lib/watcher/appmetrics.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+  UNUSED_PROBES.forEach(function(p) {
+    appmetrics.disable(p);
```
Comment (sam-github) [inline lib/watcher/appmetrics.js]: and this will break appmetrics-dash, can't be done here

Comment (kjdelisle) [inline lib/watcher/appmetrics.js]: Is this line no longer required?

**Comment by sam-github** [comment]
the problem is that this watcher assumed appmetrics was always there, and it was the only user. this isn't necessarily true. I think that the watcher will somehow have to ask the adapter if it is started, and if it is, then send the metrics up, or that the the metrics emission will have to occur from lib/adatper.js, and only the relaying up will happen in the watcher. This is the kind of complexity that made me hope that it could be done in wlpn-cli-server, but you found it can't.

File: lib/watcher/appmetrics.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  'redis',
+  'memcached',
+  'oracledb',
+  'strong-oracle',
+  'requests',
+  'trace',
+];
+
+exports.worker = function(handle) {
+  if (!global.APPMETRICS) {
```
Comment (kjdelisle) [inline lib/watcher/appmetrics.js]: I noticed that the adapter creates `global.APPMETRICS`. I'm not necessarily opposed to finding another way, but would this be sufficient for disabling this watcher if appmetrics isn't present?

File: lib/adapter.js [hunk 4]
```diff
@@ -214,10 +214,13 @@ function start(options) {
...
     appmetrics.enable('eventloop');
     monitor = appmetrics.monitor();
+
+    global.APPMETRICS = appmetrics;
+    global.APPMETRICS_MONITOR = monitor;
+
     initialise();
     appmetricsStarted = true;
 
-    global.APPMETRICS = appmetrics;
```
Comment (sam-github) [inline lib/adapter.js]: why did you move the code?

File: lib/watcher/appmetrics.js (hunk 3 above)
Comment (sam-github) [inline lib/watcher/appmetrics.js]: I think so, though its a worrying order dependency between adapter start and watcher initialization. Not sure what else to do, though.

File: lib/adapter.js (hunk 4 above)
Comment (kjdelisle) [inline lib/adapter.js]: I thought the `initialise()` function call would spin up the watchers before that definition was set?

Comment (sam-github) [inline lib/adapter.js]: that's a good reason :-)

**Comment by kjdelisle** [comment]
@slnode test please
getting errors about not being able to find `stats-lite` (but it passes on Travis CI...)

**Comment by sam-github** [comment]
What is `stats-lite`? Its not a dep of strong-supervisor, even indirectly.

Can you test manually with `sl-run --no-profile ..`, and if that passes, this LGTM (but tests do need to pass).

File: lib/watcher/appmetrics.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
+// Copyright IBM Corp. 2015,2016. All Rights Reserved.
+// Node module: strong-supervisor
+// This file is licensed under the Artistic License 2.0.
+// License text available at https://opensource.org/licenses/Artistic-2.0
+
+'use strict';
+
```
Comment (sam-github) [inline lib/watcher/appmetrics.js]: can you add a comment explaining what this is for? that it produces events used by the Liberty collective? so no one deletes it in the future on the assumption that is unused :-)

**Comment by kjdelisle** [comment]
@sam-github `stats-lite` is a devDependency of `stats-incremental`, which is a dependency of `strong-trace`.

**Comment by kjdelisle** [comment]
I can confirm that it does fix the issue, and that metrics are reaching the collective controller with this patch in place.

**Comment by rmg** [comment]
It is passing on Travis because there is no license key in the environment so the tests that are failing on cis-jenkins are being skipped there. That is also probably how it is passing for you locally.

**Comment by rmg** [comment]
The failure appears to have been introduced very recently. PR submitted upstream: brycebaril/stats-incremental#1

**Comment by rmg** [comment]
@slnode test please

**Comment by rmg** [comment]
`stats-incremental` was fixed upstream, the remaining failure looks like it _might_ be node 8 specific.

**Comment by kjdelisle** [comment]
The failures I'm seeing are somewhat racy. Travis CI is getting intermittent timeouts at various parts of the tests. Case-in-point, re-running the suite has made it go green as of this posting.
I'll give it another spin.
@slnode test please

**Comment by kjdelisle** [comment]
@slnode test please

**Comment by kjdelisle** [comment]
Tinkering with the timeout; it might need more time to finish.
For posterity, I'll mention that I'm not a fan of messing with test timeouts as it's usually a band-aid for a bigger problem, but we'll see.

**Comment by kjdelisle** [comment]
@sam-github @rmg The test passes with the timeout removed completely, but it won't work with a timeout of 360 seconds.
Is 300 seconds a hard limit, or do you have a different discrete timeout you'd be comfortable with?

**Comment by kjdelisle** [comment]
@slnode test please

File: test/test-run-appmetrics-usage.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+// This file is licensed under the Artistic License 2.0.
+// License text available at https://opensource.org/licenses/Artistic-2.0
+
+'use strict';
+
+var debug = require('./debug');
+var run = require('./run-with-ctl-channel');
+var tap = require('tap');
+
+tap.test('appmetrics are forwarded via parentCtl', function(t) {
```
Comment (rmg) [inline test/test-run-appmetrics-usage.js]: If this is the test that needs the higher timeout, try adding `{timeout: 500*1000}` (the expanded `*1000` is to indicate to readers this is in ms, not s) as a 2nd arg to `tap.test`.

Comment (kjdelisle) [inline test/test-run-appmetrics-usage.js]: I think it's the `test-runctl-clusterctl.js` test that's timing out. It also has its own custom timeout value, which might be part of it.

I'll tweak it and see if it changes anything.

**Comment by kjdelisle** [comment]
@slnode test please

**Comment by kjdelisle** [comment]
One more try, just to make sure it's not a one-off fluke.

**Comment by kjdelisle** [comment]
@slnode test please
